File: masiro.py
import json as _json
import re   as _re
import sys  as _sys

import requests as _requests

forum = None
lists = None
novel = None

auth  = None
user  = None
self  = None

class cookies:
    __files__   = []
    def __init__(self, file = None):
        self.__cookie__ = {}
        if(file):
            self.__files__.append(file)
        if(self.__files__):
            self.start(self.__files__[-1])

    def __setitem__(self, name, value):
        self.__cookie__[name] = value
        self.__handle__.write( '{}\t{}\n'.format(name, value) )

    def __getitem__(self, name):
        return self.__cookie__[name]
    
    def __del__(self):
        if( hasattr(self, '__handle__') ):
            self.__handle__.close()
    
    def start(self, file):
        with open(file, 'r') as fin:
            for line in fin.readlines():
                self.__cookie__.__setitem__( *line[:-1].split('\t', 1) )
        self.__handle__ = open(file, 'a+')
    
    def gets(self):
        return self.__cookie__

    def puts(self, get):
        for name in get.keys():
            self.__cookie__[name] = get[name]


class _executor:
    def __init__(self, config_file, use_cookie = cookies):
        self.is_load = False
        self.file = config_file
        self.rule = None
        self.cook = use_cookie

    def load(self):
        try:
            with open(self.file, 'r', encoding='UTF-8') as fin:
                self.rule = _json.load(fin)
        except FileNotFoundError:
            print("Error: File not found")
        except _json.JSONDecodeError:
            print('Error: Invalid json data in file')
        else:
            self.is_load = True

    def __call__(self, *args, **kwargs):
        if(not self.is_load):
            self.load()
        data = self.translate_variables(self.rule['data']['dataset'], args, kwargs)
        base, info = self.translate_request( data )

        response = _requests.request(*base, **info)

        if(self.rule['isSaveCookie']):
            self.cook().puts( response.cookies.get_dict() )

        return self.translate_content(response.text, self.rule['response'])
        
    
    def translate_request(self, data):
        base = ["POST"if(self.rule['isWithData'] and self.rule['data']['method'] not in ('params'))else'GET', self.rule['URL'] ]
        info = {
            'headers' : {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36',
                'Accept':'*/*',
                'Accept-Language':'zh-CN,zh;q=0.9,en;q=0.8'
            }
        }
        _sys.stderr.write( str(base)+'\n' )
        if(self.rule['isWithData']):
            info[ self.rule['data']['method'] ] = data
        if(self.rule['isWithCookie']):
            info['cookies'] = self.cook().gets()
        return base, info

    def translate_variables(self, dataset, args, kwargs):
        result = {}
        for index in range(len(args)):
            result[ dataset[index]['name'] ] = args[index]
        result.update(kwargs)
        for item in dataset:
            if(item['name'] not in result.keys()):
                if(item['required']):
                    raise AttributeError('The Attibution of {} is Required'.format(item['name']))
                if(not item["required"] and "defined" in item.keys()):
                    result[ item["name"] ] = item["defined"]
        return result

    def translate_content(self, data, method):
        result = {}
        for index in method['order']:
            if  ( method[index]['type'] == 'return' ):
                return result[ method[index]['refer'] ]

            if  ( method[index]['type'] == "regex"  ):
                if  ( method[index]['func'] == 'hunt' ):
                    get = _re.search(method[index]["rule"], data)
                    result[index] = get.group() if(get)else ''
                elif( method[index]['func'] == 'find' ):
                    result[index] = [ get.group() for get in _re.finditer(method[index]["rule"], data) ]
                elif( method[index]['func'] == 'seek' ):
                    result[index] = _re.findall(method[index]['rule'], data)
                elif( method[index]['func'] == 'exist'):
                    result[index] = _re.search(method[index]["rule"], data) != None

            elif( method[index]["type"] == 'func'   ):
                if  ( method[index]["func"] == 'json' ):
                    result[index] = _json.loads(data)
                elif( method[index]["func"] == "code" ):
                    data = data.encode().decode("unicode_escape")
                elif( method[index]["func"] == "save" ):
                    with open("temp/temp.txt", "w+", encoding='UTF-8') as fout:
                        fout.write(data)

            elif( method[index]['type'] == 'refer'  ):
                get = result
                for item in method[index]["rule"]:
                    get = get[item]
                result[index] = get

            elif( method[index]['type'] == 'range'  ):
                result[index] = []
                for sub in result[ method[index]['refer'] ]:
                    result[index].append( self.translate_content(sub, method[index]['rule']) )
            else:
                raise RuntimeError('unknown type: {}'.format(index))
            
        for name in list(result.keys()):
            if( name.startswith("temp") ):
                del result[ name ]

        return result


class _categories:
    def __init__(self, load):
        for files in load:
            setattr(self, files['name'], _executor('configure\\' + files['file']))


def __init__(config_file = 'config.json'):
    global forum, lists, novel, \
           auth,  user,  self

    with open('config.json', 'r') as fin:
        _config = _json.load(fin)

    forum = _categories( _config['forum'] )
    lists = _categories( _config['lists'] )
    novel = _categories( _config['novel'] )

    auth  = _categories( _config["auth"] )
    user  = _categories( _config['user'] )
    self  = _categories( _config['self'] )

if(__name__ == 'masiro'):
    __init__()

File: masiro.js
const axios = require("axios");
const fopen = require('fs');

axios.defaults.withCredentials = true;

var defined_cookie_file = "cookie.tmp";

const regex = {
    "hunt":(pattern, string)=>{
        let ans=string.match( RegExp(pattern) );
        if(ans)
            return ans[0];
        else   
            console.error(pattern+'\n'+string)
    },
    "seek":(pattern, string)=>{
        let ans = [],from = string.matchAll(RegExp(pattern, 'g'))
        for(let item of from)
            ans.push( item[0] );
        return ans;
    },
    "find":(pattern, string)=>{
        let ans = [],from = string.matchAll(RegExp(pattern, 'g'));
        for(let item of from)
            ans.push(item[0]);
        return ans;
    },
    "exist":(pattern, string)=>{
        return RegExp(pattern).exec(string) != null;
    }
}


function cookies(file = null){
    let cookie = {};
    let handle = null;
    if(!file){
        if(!defined_cookie_file)
            throw "Need a cookies file path";
        file = defined_cookie_file;
    }
    else{
        defined_cookie_file = file;
    }
    let data = fopen.readFileSync(file, 'utf8');
    let lines = data.split('\n');
    for (let line of lines) {
        if(line){
            const [name, value] = line.slice(0, -1).split('\t', 2);
            cookie[name] = value;
        }
    }
    handle = fopen.openSync(file, 'a+');
    return {
        "set_value":(name, value)=>{
            cookie[name] = value;
            fopen.writeSync(handle, `${name}\t${value}\n`);
        },
        "get_value":(name)=>{
            return cookie[name];
        },
        "gets":()=>{
            return cookie;
        },
        "puts":(get)=>{
            for (let name in get) 
                cookie[name] = get[name];
        },
        "encode":()=>{
            let lines = [];
            for(let name in cookie)
                lines.push(`${name}=${cookie[name]}`);
            return lines.join('; ');
        }
    }
}


async function requests(method, url, more){
    let config = more;
    config.url = url,config.method = method;
    let [text, cookies] = await axios(
        config
    ).then( response => {
        let data = response.data, cookie = response.headers['set-cookie'],real_cookie={};
        if(typeof data != "string")
            data = JSON.stringify(data);
        for(let lines of cookie){
            let line=lines.replace(/expires=.+?;/gi,'')
                          .replace(/Max-age=.+?;/gi,'')
                          .replace(/path=.+/gi,'')
                          .replace(/;/,'')
                          .trim();
            let [name, value] = line.slice(0, -1).split('=', 2);
            real_cookie[name] = value;
        }
        return [ data, real_cookie ];
    })
    return {
        "text":text,
        "cookies":cookies
    };
}


function executor(config_file, use_cookie = cookies){
    let is_load = false;
    let file = config_file;
    let rule = null;
    let cook = use_cookie;

    function load(){
        try {
            rule = JSON.parse( fopen.readFileSync(file, "utf8") );
            is_load = true;
        } catch (error) {
            console.log(error);
        }
    }

    function translate_request(data){
        let base =  [ ( rule.isWithDate && ["params"].indexOf(rule.data.method) == -1 )?"post":"get", rule.URL ];
        let info = {
            'headers' : {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36',
                'Accept':'*/*',
                'Accept-Language':'zh-CN,zh;q=0.9,en;q=0.8'
            }
        };
        if(rule.isWithData){
            switch(rule.data.method){
                case 'params':{
                    info.params = data;
                }break;
                case 'json':{
                    info.data = data;
                }break;
                default:{
                    throw "stop"
                }break;
            }
        }
        if(rule.isWithCookie){
            info.headers.cookie = cook().encode();
        }
        return [base, info];
    }
    function translate_variables(dataset, args){
        let result = {};
        for(let index in args)
            result[ dataset[index].name ] = args[index];
        for(let item of dataset){
            if( Object.keys(result).indexOf(item.name) == -1 ){
                if(item.required)
                    throw `AttributeError : The Attibution of ${item.name} is Required` ;
                if(!item.required && Object.keys(item).indexOf("defined") != -1)
                    result[ item.name ] = item.defined ;
            }
        }
        return result;
    }

    function translate_content(data, method){
        let result = {} ;
        for(let index of method.order){
            switch( method[index].type ){
                case "return":{
                    return result[ method[index].refer ]
                };

                case "regex":{
                    result[ index ] = regex[ method[index].func ]( method[index].rule, data )
                }break;

                case "func":{
                    switch (method[index].func ){
                        case "json":{
                            result[index] = JSON.parse(data);
                        }break;
                        case "code":{
                            console.error( "decode unicode" );
                        }break;
                        case "save":{
                            fopen.writeFileSync("temp/temp.txt", data, {
                                encoding:"utf8",
                                flag:"w+"
                            })
                        }break;
                        case "del":{
                            delete result[ method[index].refer ];
                        }break;
                    }
                }break

                case "refer":{
                    let get = result;
                    for(let item of method[index].rule)
                        get = get[item];
                    result[index] = get;
                }break;

                case "range":{
                    result[index] = [];
                    for(let sub of result[ method[index].refer ]){
                        result[index].push( translate_content(sub, method[index].rule) );
                    }
                }break;

                default:{
                    throw `[RuntimeError] : unknown type ${index}`;
                }
            }
        }

        console.log(result);

        for(let name in result)
            if( name.startsWith('temp') )
                delete result[ name ] ;

        return result;
    }

    return async function(...args){
        if(!is_load)
            load();
        let data = translate_variables(rule.data.dataset, args)
        let [base, info] = translate_request( data );
        let response = await requests(base[0], base[1], info);
        if(rule.isSaveCookie)
            cook().puts( response.cookies );
        return translate_content(response.text, rule.response);
    };
}


function categories(load){
    let executors = {}
    for(let index in load)
        executors[ load[index].name ] = executor('configure\\' + load[index].file);
    return executors;
}

const config = JSON.parse( fopen.readFileSync('config.json') );

const forum = categories(config.forum);
const lists = categories(config.lists);
const novel = categories(config.novel);

const auth  = categories(config.auth );
const user  = categories(config.user );
const self  = categories(config.self );

module.exports = {
    cookies,
    forum, lists, novel, 
    auth,  user,  self
};
